pagefaultgames/pokerogue
# [Feature] Implement Full Heal/Lum/Full Restore to heal Confusion

## What are the changes?

Fixes #1112 and #1054. I did not test revive last PR, despite needing to. I did not know and I apologize. This will have the testing for everything I could think of. 

## Why am I doing these changes?



To implement a feature that is already in game.

## What did change?


I changed `ressetStatus()` to include `confusion: boolean = false` to the already made parameter. It won't affect any existing `resetStatus()`, only the ones we want to heal confusion.

Besides that, I added logic to see if the the `BattlerTagType.CONFUSED` is there or not. I did that for `PokemonHPRestoreModifierType` and `PokemonStatusHealModifierType`.

I made Lum Berry use the same `resetStatus()` logic.

Lastly, I added the `resetStatus()` to `modifier.ts`. The reason it messed up revive was because I accidentally put it inside an if statement, when it should not have been.

### Screenshots/Videos


Video to show Revive works:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/52b6d716-e031-41cc-a011-15f5b292c91f)

Video to show Max Revive works:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/65122f27-779d-475a-a576-8061df3c3ba9)

Video to show Full Heal works:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/34076f90-93e5-46f9-a32c-a0399d3cb7b2)

Video to show Full Restore works at Max HP:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/a46a20d1-10e7-4a8e-9d39-549050916046)

Video to show Full Restore works at Semi Max HP:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/9db91e49-1be1-4679-abd7-316a10e709d0)

Video to show normal healing items (such as Potions) do not heal confusion:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/bc2e0b0d-bad1-4f08-9789-4384e50f7184)

Video to show Lum Berry does work in battle to heal Confusion:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/abfc7cc7-bd12-4c2d-ad44-0a710b9f7a57)

## How to test the changes?




You will need to get confused, and stay confused until the battle is over. I suggest personally to disable the chance of hurting yourself while confused during testing; it makes the whole testing period easier. Other than that: give the opposing Pokemon the prankster ability so it will use Confuse Ray first, and you should be good to go.

## Checklist
    ![image](https://github.com/pagefaultgames/pokerogue/assets/71776311/758fe714-786f-4b3f-a6de-e5065f7827e2)
